File: python/mirage/commands/builtin/ssh/ls/ls.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.accessor.ssh import SSHAccessor
from mirage.cache.index import IndexCacheStore
from mirage.commands.builtin.utils.formatting import _human_size
from mirage.commands.registry import command
from mirage.commands.spec import SPECS
from mirage.core.ssh.glob import resolve_glob
from mirage.core.ssh.readdir import readdir
from mirage.core.ssh.stat import stat
from mirage.io.types import ByteSource, IOResult
from mirage.types import FileType, PathSpec


async def _ls_entries(
    accessor: SSHAccessor,
    path: PathSpec,
    all_files: bool,
    sort_by: str,
    reverse: bool,
    recursive: bool,
    list_dir: bool,
    warnings: list[str],
    index: IndexCacheStore = None,
):
    if list_dir:
        s = await stat(accessor, path)
        return [s]

    try:
        entries = await readdir(accessor, path, index)
    except (FileNotFoundError, ValueError) as exc:
        warnings.append(f"ls: cannot access '{path.original}': {exc}")
        return []

    stats = []
    for entry in entries:
        try:
            e_spec = PathSpec(original=entry,
                              directory=entry,
                              resolved=False,
                              prefix=path.prefix)
            s = await stat(accessor, e_spec)
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{entry}': {exc}")
            continue
        if not all_files and s.name.startswith("."):
            continue
        stats.append(s)

    if sort_by == "time":
        stats.sort(key=lambda s: s.modified or "", reverse=not reverse)
    elif sort_by == "size":
        stats.sort(key=lambda s: s.size or 0, reverse=not reverse)
    else:
        stats.sort(key=lambda s: s.name, reverse=reverse)

    if recursive:
        sub_entries = []
        for s in stats:
            sub_entries.append(s)
            if s.type == FileType.DIRECTORY:
                entry_path = path.child(s.name)
                entry_spec = PathSpec(original=entry_path,
                                      directory=entry_path,
                                      resolved=False,
                                      prefix=path.prefix)
                sub = await _ls_entries(accessor, entry_spec, all_files,
                                        sort_by, reverse, recursive, False,
                                        warnings, index)
                sub_entries.extend(sub)
        return sub_entries

    return stats


@command("ls", resource="ssh", spec=SPECS["ls"])
async def ls(
    accessor: SSHAccessor,
    paths: list[PathSpec],
    *texts: str,
    stdin: bytes | None = None,
    args_l: bool = False,
    args_1: bool = False,
    a: bool = False,
    A: bool = False,
    h: bool = False,
    t: bool = False,
    S: bool = False,
    r: bool = False,
    R: bool = False,
    d: bool = False,
    F: bool = False,
    index: IndexCacheStore = None,
    cwd: PathSpec | str = "/",
    prefix: str = "",
    **_extra: object,
) -> tuple[ByteSource | None, IOResult]:
    if not paths:
        cwd_str = cwd.original if isinstance(cwd, PathSpec) else cwd
        cwd_prefix = cwd.prefix if isinstance(cwd, PathSpec) else ""
        paths = [
            PathSpec(original=cwd_str,
                     directory=cwd_str,
                     resolved=False,
                     prefix=cwd_prefix)
        ]
    paths = await resolve_glob(accessor, paths, index)
    targets = [p for p in paths]
    all_files = a or A
    sort_by = "name"
    if t:
        sort_by = "time"
    elif S:
        sort_by = "size"
    warnings: list[str] = []
    results: list[str] = []
    for p in targets:
        p.prefix if isinstance(p, PathSpec) else ""
        try:
            entries = await _ls_entries(
                accessor,
                p,
                all_files,
                sort_by,
                r,
                R,
                d,
                warnings,
                index,
            )
        except (FileNotFoundError, ValueError) as exc:
            warnings.append(f"ls: cannot access '{p.original}': {exc}")
            continue
        if args_l and not args_1:
            for e in entries:
                size_str = _human_size(e.size or 0) if h else str(e.size or 0)
                line = (f"{e.type or '-'}\t{size_str}"
                        f"\t{e.modified or ''}\t{e.name}")
                results.append(line)
        else:
            for e in entries:
                is_dir = F and e.type == FileType.DIRECTORY
                name = e.name + "/" if is_dir else e.name
                results.append(name)
    stderr = "\n".join(warnings).encode() if warnings else None
    exit_code = 1 if warnings and not results else 0
    output = "\n".join(results).encode()
    return output, IOResult(stderr=stderr, exit_code=exit_code)

File: typescript/packages/core/src/commands/builtin/langfuse/ls.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { LangfuseAccessor } from '../../../accessor/langfuse.ts'
import { resolveLangfuseGlob } from '../../../core/langfuse/glob.ts'
import { readdir as langfuseReaddir } from '../../../core/langfuse/readdir.ts'
import { stat as langfuseStat } from '../../../core/langfuse/stat.ts'
import { PathSpec, ResourceName } from '../../../types.ts'
import { command, type CommandFnResult, type CommandOpts } from '../../config.ts'
import { specOf } from '../../spec/builtins.ts'
import { lsGeneric } from '../generic/ls.ts'
import { metadataProvision } from './_provision.ts'

async function lsCommand(
  accessor: LangfuseAccessor,
  paths: PathSpec[],
  _texts: string[],
  opts: CommandOpts,
): Promise<CommandFnResult> {
  let workingPaths: PathSpec[] = paths
  if (workingPaths.length === 0) {
    workingPaths = [
      new PathSpec({
        original: opts.cwd,
        directory: opts.cwd,
        resolved: false,
        prefix: opts.mountPrefix ?? '',
      }),
    ]
  }
  const resolved = await resolveLangfuseGlob(accessor, workingPaths, opts.index ?? undefined)
  return lsGeneric(
    resolved,
    opts,
    (p) => langfuseReaddir(accessor, p, opts.index ?? undefined),
    (p) => langfuseStat(accessor, p, opts.index ?? undefined),
  )
}

export const LANGFUSE_LS = command({
  name: 'ls',
  resource: ResourceName.LANGFUSE,
  spec: specOf('ls'),
  fn: lsCommand,
  provision: metadataProvision,
})
